java/java.lsp.server/vscode/src/jdk/validation/project.ts

```diff
@@ -0,0 +1,171 @@
+import * as xml2js from 'xml2js';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as os from 'os';
+import { normalizeJavaVersion } from './javaUtil';
+
+const GET_PROJECT_INFO = 'nbls.project.info';
+const GRADLE_TARGET_COMPATIBILITY_REGEX = /targetCompatibility\s*=\s*(?:JavaVersion\.toVersion\(['"](\d+(\.\d+)?)['"]\)|['"](\d+(\.\d+)?)['"])/
+const GRADLE_SOURCE_COMPATIBILITY_REGEX = /sourceCompatibility\s*=\s*(?:JavaVersion\.toVersion\(['"](\d+(\.\d+)?)['"]\)|['"](\d+(\.\d+)?)['"])/
+
+export enum BuildSystemType {
+    MAVEN = 'Maven',
+    GRADLE = 'Gradle',
+    UNKNOWN = 'Unknown'
+}
+
+export async function getProjectFrom(projectUri: vscode.Uri): Promise<Project | undefined> {
+    const projectInfos: any[] = await vscode.commands.executeCommand(GET_PROJECT_INFO, projectUri.toString(), { projectStructure: true });
+    if (projectInfos?.length && projectInfos[0]) {
+        const projectDirectory = projectInfos[0].projectDirectory.toString();
+        const buildSystem: BuildSystemType = resolveBuildSystemType(projectUri, projectInfos[0].projectType);
+
+        switch (buildSystem) {
+            case BuildSystemType.MAVEN:
+                const mavenSubprojects: Project[] = projectInfos[0].subprojects
+                    .map((subproject: string) => new MavenProject(subproject, []))
+                return new MavenProject(projectDirectory, mavenSubprojects);
+            case BuildSystemType.GRADLE:
+                const gradleSubprojects: Project[] = projectInfos[0].subprojects
+                    .map((subproject: string) => new GradleProject(subproject, []))
+                return new GradleProject(projectDirectory, gradleSubprojects);
+            default:
+                break;
+        }
+    }
+    return Promise.resolve(undefined);
+}
+
+function resolveBuildSystemType(uri: vscode.Uri, projectType?: string): BuildSystemType {
+    if (projectType?.includes('gradle')) {
+        return BuildSystemType.GRADLE;
+    }
+    if (projectType?.includes('maven')) {
+        return BuildSystemType.MAVEN;
+    }
+    if (fs.existsSync(path.join(uri.fsPath, 'build.gradle'))) {
+        return BuildSystemType.GRADLE;
+    }
+    if (fs.existsSync(path.join(uri.fsPath, 'pom.xml'))) {
+        return BuildSystemType.MAVEN;
+    }
+    return BuildSystemType.UNKNOWN;
+}
+
+export abstract class Project {
+
+    readonly directory: string;
+    readonly subprojects: Project[];
+
+    constructor(directory: string, subprojects: any[]) {
+        this.directory = directory.replace('file:/', '');
```

Comment:
Note: this may produce `C:/Temp/whatever` paths on Windows. Possibly transform e.g. using `vscode.Uri.parse(directory).fsPath`